### DISCUSSION TITLE
mise activate not workings with --shims flag ?

### DISCUSSION THREAD

**Discussion by tony-sol** [jdx/mise#3817] [Q&A] [answered]
I've tried to use direnv-ish functional of mise and faced with strange behaviour. Previously i only used mise for tools versions management and this
```shell
source <(mise activate zsh --shims)
```
in `$ZDOTDIR/.zshrc` worked fine (as i thought so).

But while trying to use env, i figured out, that `mise doctor` will keep saying `activated: no` until i changed activation line to 
```shell
source <(mise activate zsh)
```

Next, i've tried to figure out what exactly activation does and found this: 
<img width="1110" alt="Снимок экрана 2024-12-26 в 02 47 18" src="https://github.com/user-attachments/assets/eebceb3b-2f81-499d-b086-373f96c4f051" />
so `mise activate zsh --shims` is empty. 
Okay, that's strange, but what's next - right after activate without shims-flag - there is output:
<img width="1110" alt="Снимок экрана 2024-12-26 в 02 50 04" src="https://github.com/user-attachments/assets/a37621ad-7892-4ddf-adbf-65e95b44e305" />
BUT!! - `$PATH` has been already modified while activation and another `activate --shims` will just add Yet Another path and report a problem:
```
1. shims are on PATH and mise is also activated. You should only use one of these methods
```

In my understanding, these features aren't mutually exclusive - e.g. i set up `$MISE_SHIMS` explicitly AND put in into `$PATH` explicitly as well (cause i need to keep a specific paths order), so i don't need `$PATH` to be modified, BUT i also want to have env and hooks features, provided by activation. 

So is there any recipes/hacks/workarounds to achieve that?

**Comment by tony-sol** [comment]
long and the short - is there a way to active mise, without modifying `$PATH`, except _really_ dirty hacks like this
```shell
source <(mise activate zsh) && export PATH="$(__calling_my_path_builder_function_again)"
```

**Reply by tony-sol** [reply]
I was wrong, even though `mise activate zsh` has `export PATH` in the beginning, hacks like this
```shell
source <(mise activate zsh | grep --invert-match --extended-regexp '^export PATH=')
```
don't work and if your `PATH` already has `MISE_SHIMS`, `mise doctor` would say
```
version: 2024.12.22 macos-arm64 (2024-12-30)
activated: yes
shims_on_path: yes
...
settings:

1 problem found:

1. shims are on PATH and mise is also activated. You should only use one of these methods.
```
and still activation modifies `PATH`

**Comment by syhol** [comment]
If you run activate without sourcing you can see this function that sources the result of hook-env:
```zsh
_mise_hook() {
  eval "$(/Users/simon/.local/bin/mise hook-env -s zsh)";
}
```
This is what sets the path, and this function is called in a few places.

So this is where you need to apply your ` | grep --invert-match --extended-regexp '^export PATH='`

You could save that entire output down to a file like `my-local-zsh-activate.zsh` and add the grep to that function and delete the PATH:
```zsh
_mise_hook() {
  eval "$(/Users/simon/.local/bin/mise hook-env -s zsh | grep --invert-match --extended-regexp '^export PATH=')";
}
```
Then source it instead:
```zsh
source my-local-zsh-activate.zsh
```
Or skip the separate file and edit it on the fly with sed:
```zsh
source <(mise activate zsh | grep --invert-match --extended-regexp '^export PATH=' | sed "s/hook-env -s zsh/hook-env -s zsh | grep --invert-match --extended-regexp '^export PATH='/g")
```

**Reply by tony-sol** [reply]
Hmm, i saw `mise activate zsh` output and saw `mise hook-env` there, but didn't think that there are several PATH exports
I'll research this more closely, thx

**Reply by syhol** [reply]
Some more context here: https://mise.jdx.dev/faq.html#what-does-mise-activate-do

**Comment by jdx** [comment]
> cause i need to keep a specific paths order

what does this mean? `mise activate` respects the PATH order and injects entries into the point it was activated

**Reply by tony-sol** [reply]
Here is my case for more context:
```shell
# .zshenv
export XDG_BIN_HOME="${XDG_BIN_HOME:-${HOME}/.local/bin}"
export XDG_DATA_HOME="${XDG_DATA_HOME:-${HOME}/.local/share}"
export HOMEBREW_PREFIX="/opt/homebrew"
export MAVEN_HOME="${XDG_DATA_HOME}/m2"
export M2_HOME="${MAVEN_HOME}"
export M2="${M2_HOME}/bin"
export GOPATH="${XDG_DATA_HOME}/go"
export GOBIN="${GOPATH}/bin"
export GEM_HOME="${XDG_DATA_HOME}/gem"
export GEM_BIN="${GEM_HOME}/bin"
export MISE_DATA_DIR="${XDG_DATA_HOME}/mise"
export MISE_SHIMS="${MISE_DATA_DIR}/shims"
export KREW_ROOT="${XDG_DATA_HOME}/krew"
export KREW_BIN="${KREW_ROOT}/bin"
export MASON_BIN="${XDG_DATA_HOME}/nvim/mason/bin"
export PYTHONUSERBASE="${XDG_DATA_HOME}/python"

# .zshrc
function __prepend_path {
	local path=$1
	shift
	for value in $@
	do
		if [[ ":${path}:" == *:"$value":* ]]; then
			path="${path//$value/}"
		fi
		path="${value}${path+:${path//::/:}}"
	done
	echo ${${path#:}%:}
}
export PATH=$(__prepend_path $PATH "${HOMEBREW_PREFIX}/sbin" "${HOMEBREW_PREFIX}/bin" $MISE_SHIMS $M2 $GOBIN $GEM_BIN $KREW_BIN $MASON_BIN "${PYTHONUSERBASE}/bin" $XDG_BIN_HOME)
# ...
source <(mise activate zsh)
```

So this is required to bypass macos's `/usr/libexec/path_helper` and ensure that `PATH` variable is idempotent regardless of the terminal and/or is default cmd (e.g. i have same `PATH` and inside vscode's term, and in native Terminal.app and WezTerm even with `{ 'zsh', '--login', '--interactive', '-c', 'tmux new -A' }` as default_prog)

"i need to keep a specific paths order" means, that i want a full control of how my cli tools called. 
E.g. homebrew's bin need to be over default `/usr/bin:/usr/sbin` due to prevent default tool been called when there is homebrew's one - like native system git and git installed via brew.
Next, i want to have mise shims over homebrew's bin for the same reason - python installed via mise (it was `python = "system"` until `"system"` became deprecated), must be over homebrew's one, which is over native one.
And finally i want to have `GOBIN`, `GEM_BIN`, `XDG_BIN_HOME` (which is not in xdg-spec, i know, but it's actually used by some tools) and etc. for tools, which i build explicitly.

So, if you say that
> mise activate respects the PATH order and injects entries into the point it was activated

as i understand this, it means i should split my `PATH` exporting on two parts: before place i want to injects mise entries, next call `source <(mise activate)` to inject, and next add the rest entries, like

```shell
export PATH=$(__prepend_path $PATH "${HOMEBREW_PREFIX}/sbin" "${HOMEBREW_PREFIX}/bin")
source <(mise activate zsh)
export PATH=$(__prepend_path $PATH $M2 $GOBIN $GEM_BIN $KREW_BIN $MASON_BIN "${PYTHONUSERBASE}/bin" $XDG_BIN_HOME)
```
and, imho, it's really dirty workaround

**Reply by jdx** [reply]
I could be wrong but I think mise will also inject to where the shims directory was if that was previously in PATH. So you could add the shims, then call `mise activate` which will remove the shims but inject mise entries where it was.

**Reply by tony-sol** [reply]
And (i think) the most important part is there is no way to enable mise's hooks (those ones which are used for `[env]` or `[hooks]` sections) **with shims**:

`mise activate zsh`, no shims in $PATH gives:
```shell
export MISE_SHELL=zsh
export __MISE_ORIG_PATH="$PATH"

mise() {
  local command
  command="${1:-}"
  if [ "$#" = 0 ]; then
    command /opt/homebrew/bin/mise
    return
  fi
  shift

  case "$command" in
  deactivate|shell|sh)
    # if argv doesn't contains -h,--help
    if [[ ! " $@ " =~ " --help " ]] && [[ ! " $@ " =~ " -h " ]]; then
      eval "$(command /opt/homebrew/bin/mise "$command" "$@")"
      return $?
    fi
    ;;
  esac
  command /opt/homebrew/bin/mise "$command" "$@"
}

_mise_hook() {
  eval "$(/opt/homebrew/bin/mise hook-env -s zsh)";
}
typeset -ag precmd_functions;
if [[ -z "${precmd_functions[(r)_mise_hook]+1}" ]]; then
  precmd_functions=( _mise_hook ${precmd_functions[@]} )
fi
typeset -ag chpwd_functions;
if [[ -z "${chpwd_functions[(r)_mise_hook]+1}" ]]; then
  chpwd_functions=( _mise_hook ${chpwd_functions[@]} )
fi

_mise_hook
if [ -z "${_mise_cmd_not_found:-}" ]; then
    _mise_cmd_not_found=1
    [ -n "$(declare -f command_not_found_handler)" ] && eval "${$(declare -f command_not_found_handler)/command_not_found_handler/_command_not_found_handler}"

    function command_not_found_handler() {
        if /opt/homebrew/bin/mise hook-not-found -s zsh -- "$1"; then
          _mise_hook
          "$@"
        elif [ -n "$(declare -f _command_not_found_handler)" ]; then
            _command_not_found_handler "$@"
        else
            echo "zsh: command not found: $1" >&2
            return 127
        fi
    }
fi
```

`mise activate zsh --shims`, no shims in path gives:
```shell
export PATH="/Users/tony/.local/share/mise/shims:$PATH"
```
looks like this is not an "activation" at all

**Reply by tony-sol** [reply]
>  I think mise will also inject to where the shims directory was if that was previously in PATH

Just tried, no, it's not - it deletes shims and **appends** entries to PATH, so these entries are always on top

**Reply by jdx** [reply]
> And (i think) the most important part is there is no way to enable mise's hooks with shims

well this is impossible, shims can't run when you do something like enter a directory

> Just tried, no, it's not - it deletes shims and appends entries to PATH, so these entries are always on top

hmm well perhaps we should change it so that is what happens

**Reply by jdx** [reply]
or hmm, maybe it's not a good idea to do that. If someone sets shims in their login shell then later calls activate, mise not put itself early enough to override things it should.

**Reply by jdx** [reply]
I think what you should do is have a function in your zshrc used like this instead:

```zsh
__add_to_path "${HOMEBREW_PREFIX}/sbin"
__add_to_path "${HOMEBREW_PREFIX}/bin"
eval "$(mise activate zsh)"
__add_to_path "$GOBIN"
# ...
```

basically like fish's [add_to_path](https://fishshell.com/docs/current/cmds/fish_add_path.html). Putting them on separate lines will not only make the mise activation more natural but it'll make source control diffs better as well.

**Reply by tony-sol** [reply]
>> And (i think) the most important part is there is no way to enable mise's hooks with shims
>
> well this is impossible, shims can't run when you do something like enter a directory

i don't really understand how these thing are connected.
I mean, how sourcing hooks and `typeset -ag chpwd_functions;` and stuff are connected to PATH modification and why this is impossible to source hooks without modifying a PATH

**Reply by tony-sol** [reply]
> I think what you should do is have a function in your zshrc used like this instead:
> 
> ```shell
> __add_to_path "${HOMEBREW_PREFIX}/sbin"
> __add_to_path "${HOMEBREW_PREFIX}/bin"
> eval "$(mise activate zsh)"
> __add_to_path "$GOBIN"
> # ...
> ```
> 
> basically like fish's [add_to_path](https://fishshell.com/docs/current/cmds/fish_add_path.html). Putting them on separate lines will not only make the mise activation more natural but it'll make source control diffs better as well.

Yeah, i wrote about this workaround above, it looks girty for me, but works as expected
I still hope there will be native solution some day

UPD: this is how i managed this for now - https://github.com/tony-sol/.config/commit/9a40204930f0930a3ed9b3e1961cf9f472b95431

**Reply by jdx** [reply]
my point is that I think it's dirty only because you're trying to put it all on a single line, I think if you put it on multiple it won't look so bad. Of course this is cosmetic and subjective though.

> i don't really understand how these thing are connected.

because it's documented that `mise activate --shims` only adds the shim directory to PATH and nothing more. There are also technical reasons like you need to be able to put it into your login shell or possibly zshenv and mise has to be super simple. Shims can be thought of as a "compatibility mode" where not much can happen. If someone used `mise activate --shims` and mise put in chpwd logic that executes hooks—inside of a zshenv script it may very well call mise again which can call zsh and cause a loop.

**Reply by syhol** [reply]
I wouldn't consider splitting your prepend_path to be dirty. Also shims are arguably slightly less reliable than activate, so if you have the option to avoid it then it's another win.

**Reply by tony-sol** [reply]
> because it's documented that mise activate --shims only adds the shim directory to PATH and nothing more

i think i got the point, but this seems to me not quite obvious. I mean it seems something wrong in naming this way, imho after your explanation, `mise shims` would be more clear and transparent

**Reply by syhol** [reply]
I'll update the docs to more clear about this behavior

**Reply by tony-sol** [reply]
i don't think this dirty because i'm "trying to put it all on a single line", but because this hack implies multiple exporting of same variable, which increases first command lag and first prompt lag a bit.
Not dramatically, but still
Between this
```
# ...
__append_to $PATH "${HOMEBREW_PREFIX}/bin"
__append_to $PATH "${HOMEBREW_PREFIX}/sbin"
source <(mise activate)
__append_to $PATH $GEM_BIN
__append_to $PATH $GOBIN
__append_to $PATH $XDG_BIN_HOME
# ...
```
and this
```
__append_to $PATH "${HOMEBREW_PREFIX}/bin" "${HOMEBREW_PREFIX}/sbin"
source <(mise activate)
__append_to $PATH $GEM_BIN $GOBIN $XDG_BIN_HOME
```
i'd choose second one for the good of performance despite the fact that first one is more readable and looks more pretty (and i completely agree with that).
There is somewhere a line between code beauty and it's performance and mine is here :)

Thank you all for helping <3

**Comment by tony-sol** [comment, marked as answer]
Resume:

- `source <(mise activate zsh --shims)` only adds the shim directory to PATH and nothing more, if you add MISE_SHIMS into PATH manually, there is no need to call activate with **--shims** flag
- `source <(mise activate zsh)` adds its entries at the begining of PATH, if you need full control over entries order in your PATH, you should call `source <(mise activate zsh)` in place where your PATH built
